Skeleton view of the first screen of the Edit Coupon flow

pmusolino opened the pull request:
Part of #6488 

### Description
As part of the editing flow for editing a coupon, I started implementing the skeleton view of the main view with all the elements but without any logic behind it for the moment.

Note: added @toupper as a reviewer as part of his onboarding process.

### Testing instructions
1. Launch the app with the new feature flag `couponEditing` enabled.
2. Navigate to a coupon detail.
3. Press the ellipsis button in the navigation bar.
4. You should be able to see the new action "Edit Coupon".
5. Press the "Edit Coupon" action -> The new screen should be presented. 

### Demo

https://user-images.githubusercontent.com/495617/161095601-e125c95f-dbef-4164-9ca4-ea64df255795.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

pmusolino committed:
fix: navigation bar appearance

pmusolino committed:
feat: added header in AddEditCoupon

pmusolino committed:
fix: localized string

pmusolino committed:
feat: alignment in TitleAndTextFieldRow

pmusolino committed:
feat: Coupon Details section

pmusolino committed:
feat: progress on AddEditCoupon view

pmusolino committed:
fix: button style

pmusolino committed:
feat: coupon expiry date field

pmusolino committed:
feat: toggle row include free shipping

pmusolino committed:
feat: start the implementation for applying the coupon to specific products or categories

pmusolino committed:
feat: improvements to apply coupon to section

pmusolino committed:
refactor: better organization of views in Groups

pmusolino committed:
fix: added comment todo

pmusolino committed:
Merge branch 'trunk' into issue/6488-skeleton-view-for-editing-coupon

toupper commented:
Nice job @pmusolino! 

After testing, I have a couple of questions :)
- When testing on iPhone 13 simulator, the title is truncated by tail as you can see in this screenshot. Is this correct?

![Annotation on 2022-04-04 at 12-02-42](https://user-images.githubusercontent.com/1864060/161521791-da8fd66f-4e45-4bd6-ac10-0d6b90135fb7.png)

- On design we have a button "All Products" under "APPLY THIS COUPON TO", but it is missing when I test it. Is it because "All Products" is already selected and thus it doesn't show?

![Annotation on 2022-04-04 at 12-06-23](https://user-images.githubusercontent.com/1864060/161522482-9644b03e-f2e1-4286-8037-6ac0ede681ca.png)

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:
```diff
@@ -15,29 +15,196 @@ struct AddEditCoupon: View {
-                                    }
-                                }
+                            Group {
+                                TitleAndTextFieldRow(title: Localization.couponAmountPercentage,
+                                                     placeholder: Localization.couponAmountPercentage,
+                                                     text: $viewModel.amountField,
+                                                     editable: false,
+                                                     fieldAlignment: .leading,
+                                                     keyboardType: .asciiCapableNumberPad)
```
I imagine this PR only focuses on the UI, so maybe the texts are not finalized. Just want to point out that the title for the Amount may need to consider the coupon discount type, so that we can display either "Amount (%)" and "Amount (currency symbol)" depending on the type.

Regarding the placeholder, I think we should display either a potential value (e.g "0" in this case), or a message (e.g "enter amount"). Duplicating the title makes it look a bit redundant - WDYT?

Also - the field should be editable right?

And the keyboard type should be `.decimalPad` since we may want to be able to enter floating number as well.

pmusolino replied:
> I imagine this PR only focuses on the UI, so maybe the texts are not finalized. Just want to point out that the title for the Amount may need to consider the coupon discount type, so that we can display either "Amount (%)" and "Amount (currency symbol)" depending on the type.

Absolutely yes, It's something I'm handling in a separate PR 👍 

> Regarding the placeholder, I think we should display either a potential value (e.g "0" in this case), or a message (e.g "enter amount"). Duplicating the title makes it look a bit redundant - WDYT?

Same, I'm handling the placeholder in a separate PR

> Duplicating the title makes it look a bit redundant

I agree, it was just a placeholder of placeholder 🤣 

> Also - the field should be editable right?
> And the keyboard type should be .decimalPad since we may want to be able to enter floating number as well.

Yes I'm handling all these small details in a separate PR, sorry for not saying this in the PR description 🙇

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:
```diff
@@ -15,29 +15,196 @@ struct AddEditCoupon: View {
+                                .padding(.horizontal, Constants.margin)
+
+                            Group {
+                                TitleAndTextFieldRow(title: Localization.couponCode,
+                                                     placeholder: Localization.couponCode,
+                                                     text: $viewModel.codeField,
+                                                     editable: false,
+                                                     fieldAlignment: .leading,
+                                                     keyboardType: .asciiCapableNumberPad)
```
Same comment as above - please reconsider the placeholder, editable and keyboard type for the row. I'm thinking we only allow alpha-numeric characters, so maybe keyboard type `.namePhonePad` is more suitable?

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:
```diff
@@ -15,29 +15,196 @@ struct AddEditCoupon: View {
+                                .padding(.horizontal, Constants.margin)
+
+                            //TODO: leading aligning for this button
+                            Button {
+                                //TODO: handle action
+                            } label: {
+                                Text(Localization.regenerateCouponCodeButton)
+                            }
+                            .buttonStyle(LinkButtonStyle())
+                            .padding(.horizontal, Constants.margin)
```
I think the button is left-aligned in the design - is it possible to make it align to the left instead of center?

<img width="300" alt="Screen Shot 2022-04-04 at 17 28 03" src="https://user-images.githubusercontent.com/5533851/161525733-919fe0d3-d822-4661-bd84-07571591342f.png">

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:
```diff
@@ -15,29 +15,196 @@ struct AddEditCoupon: View {
+                            // TODO: add a new style with the icon on the left side
+                            Button {
+                                //TODO: handle action
+                            } label: {
+                                Text(Localization.editProductsButton)
+                                    .bodyStyle()
+                            }
```
I think the button is missing the pencil icon here 😅

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:
```diff
@@ -15,29 +15,196 @@ struct AddEditCoupon: View {
+                            // TODO: add a new style with the icon on the left side
+                            Button {
+                                //TODO: handle action
+                            } label: {
+                                Text(Localization.editProductCategoriesButton)
+                                    .bodyStyle()
+                            }
```
Same comment as above - the button should have a pencil icon on the left.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift:
```diff
@@ -15,29 +15,196 @@ struct AddEditCoupon: View {
+                        .padding(.top, Constants.verticalSpacing)
+                    }
+                }
+                .ignoresSafeArea(.container, edges: [.horizontal, .bottom])
```
I think maybe we should keep the content above the bottom safe area? Just so the action button at the bottom will not be too close to the bottom edge of the screen:

<img src="https://user-images.githubusercontent.com/5533851/161526593-914a28fc-3198-4ead-9be5-d92512a84c75.png" width=320 />

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift:
```diff
@@ -22,7 +22,16 @@ final class AddEditCouponViewModel: ObservableObject {
-    @Published private(set) var coupon: Coupon?
+    private var coupon: Coupon? {
+        didSet {
+            amountField = coupon?.amount ?? ""
+            codeField = coupon?.code ?? ""
```
Nit: How about extracting this to a method to populate data - we may need to add more logic to it such as accepted products, categories and expiry date.

pmusolino replied:
Sure it's a good idea, I'll do that when I'll start implementing the core logic 👍

itsmeichigo reviewed (commented):
Great work @pmusolino! I spot a few issues with my eagle eyes though:

- There's a typo in the title for the fixed cart discount (not fixed card):
<img src="https://user-images.githubusercontent.com/5533851/161527247-0542cca5-948e-4d49-aefe-8a3e96e99779.png" width=320 />

- The content should stay inside the safe area in the landscape mode:
<img width="806" alt="Screen Shot 2022-04-04 at 17 41 31" src="https://user-images.githubusercontent.com/5533851/161527717-5b937bcc-58b9-46ea-b2af-0c568a567770.png">

- One last (minor) issue: should we be more mindful about the vertical spacing between each element on the screen? I see a few spaces between the section headers/rows/dividers that don't quite match with the design. Usually I prefer not using the default spacing from the `VStack`s, so I'd use zero spacing and set my own spacing instead. Would that be nicer or a bit too much hassle? I think it's worth it to keep a more consistent layout 😅.

pmusolino commented:
@toupper 

> When testing on iPhone 13 simulator, the title is truncated by tail as you can see in this screenshot. Is this correct?

This is the default behavior of the `NavigationView`, btw we should check, ideally at the end of the implementation of the view, if we want to put the title outside of the navigation view. For me it's not ideal, but it's something that we can discuss later.

> On design we have a button "All Products" under "APPLY THIS COUPON TO", but it is missing when I test it. Is it because "All Products" is already selected and thus it doesn't show?

We discussed the proposal for removing the "All products" button [here](https://a8c.slack.com/archives/C02TD63FZA8/p1648739568229449). 👍

pmusolino commented:
@itsmeichigo 

> There's a typo in the title for the fixed cart discount (not fixed card)

I fixed it in the current PR on which I'm working now. 🙇 

> The content should stay inside the safe area in the landscape mode:

This is something that I'm handling in a new PR since I changed something on the UI 😄

> One last (minor) issue: should we be more mindful about the vertical spacing between each element on the screen? I see a few spaces between the section headers/rows/dividers that don't quite match with the design. Usually I prefer not using the default spacing from the VStacks, so I'd use zero spacing and set my own spacing instead. Would that be nicer or a bit too much hassle? I think it's worth it to keep a more consistent layout

Since in the last PR on which I'm working on I did some changes to the design, I'll fix that on the new PR 👍

pmusolino commented:
Merging this right now, since another PR is coming with most of the fixes requested here 👍
